src/bindings/rust/elektra/src/keyset.rs

```diff
@@ -0,0 +1,595 @@
+pub type Cursor = elektra_sys::cursor_t;
+
+bitflags! {
+    /// Bitflags to be passed to [`lookup`](struct.KeySet.html#method.lookup) and [`lookup_by_name`](struct.KeySet.html#method.lookup_by_name).
+    #[derive(Default)]
+    pub struct LookupOption: elektra_sys::option_t {
+        /// No Option set.
+        const KDB_O_NONE = elektra_sys::KDB_O_NONE as elektra_sys::option_t;
+        /// The found key will be popped from the keyset.
+        const KDB_O_POP = elektra_sys::KDB_O_POP as elektra_sys::option_t;
+        /// Ignore case.
+        const KDB_O_NOCASE = elektra_sys::KDB_O_NOCASE as elektra_sys::option_t;
+        /// Search with owner.
+        const KDB_O_WITHOWNER = elektra_sys::KDB_O_WITHOWNER as elektra_sys::option_t;
+        /// Linear search from start -> cursor to cursor -> end.
+        const KDB_O_NOALL = elektra_sys::KDB_O_NOALL as elektra_sys::option_t;
+    }
+}
+
+impl Drop for KeySet {
+    fn drop(&mut self) {
+        unsafe {
+            elektra_sys::ksDel(self.as_ptr());
+        }
+    }
+}
+
+impl Default for KeySet {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AsRef<elektra_sys::KeySet> for KeySet {
+    fn as_ref(&self) -> &elektra_sys::KeySet {
+        unsafe { self.ptr.as_ref() }
+    }
+}
+
+impl KeySet {
+    /// Returns the raw pointer of the KeySet.
+    /// Should be used with caution. In particular,
+    /// the pointer should only be modified with
+    /// `elektra_sys::ks*` functions, but `ksDel`
+    /// should not be called.
+    pub fn as_ptr(&mut self) -> *mut elektra_sys::KeySet {
+        self.ptr.as_ptr()
+    }
+
+    /// Create a new empty KeySet.
+    ///
+    /// # Panics
+    /// Panics if an allocation error (out of memory) occurs in the C-constructor.
+    pub fn new() -> Self {
+        let ks_ptr = unsafe { elektra_sys::ksNew(0, elektra_sys::KEY_END) };
+        unsafe { KeySet::from_ptr(ks_ptr) }
+    }
+
+    /// Create a new KeySet that allocates enough space for the
+    /// given capacity of keys.
+    ///
+    /// # Panics
+    /// Panics if an allocation error (out of memory) occurs in the C-constructor.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let ks_ptr = unsafe { elektra_sys::ksNew(capacity, elektra_sys::KEY_END) };
+        unsafe { Self::from_ptr(ks_ptr) }
+    }
+
+    /// Construct a new KeySet from a raw KeySet pointer
+    ///
+    /// # Panics
+    /// Panics if the provided pointer is null.
+    unsafe fn from_ptr(keyset_ptr: *mut elektra_sys::KeySet) -> KeySet {
+        KeySet {
+            ptr: std::ptr::NonNull::new(keyset_ptr).unwrap(),
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Append a KeySet to self.
+    pub fn append(&mut self, to_append: &KeySet) -> isize {
+        unsafe { elektra_sys::ksAppend(self.as_ptr(), to_append.as_ref()) }
+    }
+
+    /// Append a key to the keyset.
+    /// Returns an InsertionError if TODO: when exactly?
```

Comment:
The documentation states that return value is -1 "if insertion failed". Can we describe this in more detail when exactly insertion of a key fails?

Reply:
I created #2946 to improve the docu (and fix a possible bug). Is it clear there?

Reply:
Yes, thank you! Since it only fails on memory problems I can actually remove the error completely and panic instead to be consistent with what happens in the rest of the library.

Reply:
Yes, sounds good.